src/build_cfg.c

```diff
@@ -72,24 +72,34 @@ static void add_instruction(
     struct Location location,
     enum CfInstructionKind k,
     const union CfInstructionData *dat,
-    int noperands,
-    const struct CfVariable **operands,
+    const struct CfVariable **operands, // NULL terminated, or NULL for empty
     const struct CfVariable *destvar)
 {
-    struct CfInstruction ins = { .location=location, .kind=k, .noperands=noperands, .destvar=destvar };
+    struct CfInstruction ins = { .location=location, .kind=k, .destvar=destvar };
     if (dat)
         ins.data=*dat;
 
-    if (noperands) {
-        size_t nbytes = sizeof(ins.operands[0]) * noperands;  // NOLINT
+    while (operands && operands[ins.noperands])
+        ins.noperands++;
+    if (ins.noperands) {
+        size_t nbytes = sizeof(ins.operands[0]) * ins.noperands;  // NOLINT
         ins.operands = malloc(nbytes);
         memcpy(ins.operands, operands, nbytes);
     }
 
     Append(&st->current_block->instructions, ins);
 }
 
-#define add_varcpy(st, loc, from, to) add_instruction((st), (loc), CF_VARCPY, NULL, 1, (const struct CfVariable*[]){(from)}, (to))
+// add_instruction() takes many arguments. Let's hide the mess a bit.
+#define add_unary_op(st, loc, op, arg, target) \
+    add_instruction((st), (loc), (op), NULL, (const struct CfVariable*[]){(arg),NULL}, (target))
+#define add_binary_op(st, loc, op, lhs, rhs, target) \
+    add_instruction((st), (loc), (op), NULL, (const struct CfVariable*[]){(lhs),(rhs),NULL}, (target))
+#define add_constant(st, loc, c, target) \
+    add_instruction((st), (loc), CF_CONSTANT, &(union CfInstructionData){ .constant=copy_constant(&(c)) }, NULL, (target))
+#define add_store(st, loc, ptr, value) \
+    add_instruction((st), (loc), CF_STORE_TO_POINTER, NULL, (const struct CfVariable*[]){(ptr),(value),NULL}, NULL)
```

Comment:
Not sure if I should delete `add_store()`, or redefine it as:

```suggestion
    add_binary_op((st), (loc), CF_STORE_TO_POINTER, (ptr), (value), NULL)
```